### PR TITLE
RNG and entropy improvements

### PR DESCRIPTION
**Pull request by hannesm**
See individual commits for further details. This PR is supposed to improve the RNG and entropy story for MirageOS (adapting the FreeBSD design)
- in high load scenarios (lots of events), do not call rdseed/rdrand on every event loop iteration
- limit the frequency of reseeding (1 second)
- only reseed if a reasonable amount of entropy has been collected (64 bytes, as used by FreeBSD)
- use rdrand in favour of rdseed during "normal operation" (every second, in a timer loop)

I'd be happy if anyone is interested in looking into more detail here (keen to answer questions if you have any, maybe @cfcs is interested or others). The code as-is is ready for MirageOS, I'd shuffle it slightly around (into mirage-crypto-rng.mirage) and develop a separate mirage-crypto-rng.lwt for entropy collection with lwt (in the same vein).

### CODE REVIEW

**Comment by hannesm** [comment]
this PR is now ready. to summarize, entropy collection:
- unix uses getrandom(), hardware-assisted CPU rng, whirlwind for seeding
- lwt uses the above for seeding, also periodic (1s) rdrand for each pool, and periodic (10s) getrandom into each pool (one syscall, distribute random over pools), plus timer
- mirage uses whirlwind and hardware-assisted cpu rng for seeding, and periodic (1s) rdrand feed into each pool
--> the rng is no longer `fork()`-safe (this is a down-side, but it is faster now, and no longer amplifies high loads)

instead of a `mirage-crypto-entropy` opam package, everything is now in mirage-crypto-rng, which has three sublibraries: unix (as before), lwt (new), mirage (earlier known as mirage-crypto-entropy).

with this design, virtio-rnd can be easily integrated (though the `sources` is hardcoded and needs to be extended at a central place -- which makes sense since they need unique ids).

failure/initialization semantics: due to intricate entropy collection and RNG initialization, by default the default generator is not set. set_default_generator raises on second call (not sure anymore about this -- its meant to avoid people shooting themselves into their feet (installing a generator, setting up entropy collection, and then installing another generator)), but it may be inconvenient in the current ecosystem where some libraries call Mirage_crypto_rng_unix.initialize (tls / dns). in the mirage setting, this should be fine since the mirage tools injects the initialization call.

File: rng/fortuna.ml [hunk 1]
```diff
@@ -61,54 +73,48 @@ let generate_rekey ~g bytes =
...
+        let now = f () in
+        Int64.(sub now g.last_reseed > min_time_duration), now
+    in
+    if should_reseed then begin
+      g.reseed_count <- g.reseed_count + 1;
+      g.last_reseed <- now;
+      g.pool0_size <- 0;
+      reseedi ~g @@ fun add ->
+      for i = 0 to 31 do
+        if g.reseed_count land (1 lsl i - 1) = 0 then
```
[the remark this answers is not in the document]
Comment (hannesm) [reply rng/fortuna.ml]: done in c1e7b5c

File: rng/entropy.ml [hunk 2]
```diff
@@ -30,113 +30,128 @@
...
+
+let add_source s = _sources := s :: !_sources
+
+let sources () = !_sources
+
+let cpu_rng = function
+  | `Rdseed -> Cpu_native.rdseed
+  | `Rdrand -> Cpu_native.rdrand
+
+let random prefered =
```
[the remark this answers is not in the document]
Comment (hannesm) [reply rng/entropy.ml]: fixed in c1e7b5c

File: rng/fortuna.ml [hunk 3]
```diff
@@ -61,54 +73,48 @@ let generate_rekey ~g bytes =
...
+      | Some f ->
+        let now = f () in
+        Int64.(sub now g.last_reseed > min_time_duration), now
+    in
+    if should_reseed then begin
+      g.reseed_count <- g.reseed_count + 1;
+      g.last_reseed <- now;
+      g.pool0_size <- 0;
+      reseedi ~g @@ fun add ->
+      for i = 0 to 31 do
```
[the remark this answers is not in the document]
Comment (hannesm) [reply rng/fortuna.ml]: done in c1e7b5c (using `pools` defined at the top of the module; also use `pools` for the `land` further down)

File: rng/entropy.ml [hunk 4]
```diff
@@ -30,113 +30,128 @@
...
+  | xs when List.mem prefered xs -> Some prefered
+  | y::_ -> Some y
+
+let write_header source data =
+  Cstruct.set_uint8 data 0 source;
+  Cstruct.set_uint8 data 1 (Cstruct.len data - 2)
+
+let header source data =
+  let hdr = Cstruct.create 2 in
+  let buf = Cstruct.append hdr data in
```
[the remark this answers is not in the document]
Comment (hannesm) [reply rng/entropy.ml:90]: I find your proposed code harder to read. I don't think this makes a difference at the moment (looking at the use sides of `header`), and also think that the whole "we allocate lots of cstruct.t" needs to be revised throughout the MirageOS ecosystem.

File: rng/lwt/mirage_crypto_rng_lwt.ml [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    Logs.warn
+      (fun m -> m "Mirage_crypto_rng_lwt.initialize was called before, you \
+                   should ensure this call is intentional.")
+  else
+    (try
+       let _ = default_generator () in
+       Logs.warn (fun m -> m "Mirage_crypto_rng.default_generator has already \
+                              been set, check that this call is intentional");
+     with
+       No_default_generator -> ());
```
[the remark this answers is not in the document]
Comment (hannesm) [reply rng/lwt/mirage_crypto_rng_lwt.ml:46]: no, I explicitly want to retain two different messages here to be able to analyze the different conditions.

**Comment by hannesm** [comment]
thanks for your review @cfcs. to answer the remaining questions:

> Why is it that we cannot just auto-seed from `getrandom` on Unix on library initialization?

If I understand you correctly, you propose the `Mirage_crypto_rng_unix.initialize` being called at toplevel, and thus applications get a seeded default RNG? Unfortunately this does not work:
- the application still needs to depend on `mirage-crypto-rng.unix` to get the code linked
- only if `-linkall` is passed to ocamlopt, the Mirage_crypto_rng_unix top-level statements are evaluated, since nothing else is used by the user application
- if Lwt is used in the application, it's better (entropy collection, periodic reseeding) to use `Mirage_crypto_rng_lwt.initialize` in the Lwt event loop

> (But why is this allowed on lwt when it's disallowed on mirage? The semantic inconsistency seems unnecessary)

- On MirageOS, the `mirage` tool emits when you use the Mirage_random.S a call to `Mirage_crypto_rng_mirage.initialize ()` as part of its generated `main.ml`. This means here we can be strict and require this to be the only initialization call.
- Historical reasons: at the moment, some libraries (at least: [tls](https://github.com/mirleft/ocaml-tls/blob/v0.12.0/lwt/tls_lwt.ml#L264-L265), [dns-client.lwt](https://github.com/mirage/ocaml-dns/blob/v4.5.0/lwt/client/dns_client_lwt.ml#L21), [dns-client.unix](https://github.com/mirage/ocaml-dns/blob/v4.5.0/unix/client/dns_client_unix.ml#L23), maybe more) emit calls to `Mirage_crypto_rng_unix.initialize ()`, and I'd like to keep the ecosystem in a working state (multiple calls to `Mirage_crypto_rng_unix.initialize` are harmless~~, but once the preferable `Mirage_crypto_rng_lwt.initialize` is used, nobody should call `Mirage_crypto_rng_unix.initialize` since that will disable the entropy collection for the default rng)~~. I'm convinced these calls to RNG initialization from library code should go away (and move to the application).

Thinking again about the topic, I pushed a8c7bbd which let's the entropy collection in Mirage_crypto_rng_lwt always feed the `default_generator`; thus the entropy will reach the right generator (i.e. a call to Mirage_crypto_rng_unix.initialize after Mirage_crypto_rng_lwt.initialize is mostly harmless). This will allow for a smooth migration. Once there aren't any libraries calling Mirage_crypto_rng_unix.initialize, we can move the warnings in initialization to errors (and so unify the semantics between unix/lwt/mirage). WDYT?

**Comment by cfcs** [comment]
@hannesm Thanks for your answers, I think both make sense, and agree with the way forward for `initialize` (and that libraries should not have to muck about with the rng initialization independently).

**Pull request merged**